### DISCUSSION TITLE
表单中加载的项为自定义组件，在自定义组件中使用setvalues，不生效

### DISCUSSION THREAD

**Discussion by qreal829475** [alibaba/formily#2735] [Q&A 2.x]
onClick={(e) => {
              let currentValue = e.currentTarget.dataset;
              form.setValues({
                adtype: currentValue
              },  'merge');
              console.log(props);
            }}
在组件的click事件中，使用setvalues，但是表单中未生效。页面也也未进行重新渲染

https://codesandbox.io/s/formily-json-demo-new-g9tdr?file=/App.tsx
在例子中的63行使用的，求助，为什么会有这种不生效的情况

**Comment by qreal829475** [comment]
我发现问题了，是初始化是字符串，但是在setvalues的时候传错了，写成了对象，更新失败。
但是同样的数据，使用props中的onchange可以修改成功。

尝试onchange以后发现了是类型的问题，请问这种修改form中的值，使用setvalues和onchange哪种更好呢？

**Reply by janryWang** [reply]
没有更好一说，不同场景推荐不同方案，setValues适合在业务顶层逻辑控制，不推荐在自定义组件内部使用，onChange是自定义组件的默认约定

**Reply by qreal829475** [reply]
> 没有更好一说，不同场景推荐不同方案，setValues适合在业务顶层逻辑控制，不推荐在自定义组件内部使用，onChange是自定义组件的默认约定

还有一个问题请教一下，在ArrayItems中，每项是一个子组件，但是在子组件中怎么获取到当前的index呢？

**Reply by janryWang** [reply]
ArrayItems.useIndex()
